src/geom/sides.rs

```diff
@@ -15,6 +15,111 @@ pub struct Sides<T> {
     pub bottom: T,
 }
 
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq, Hash)]
+pub struct Margin {
+    pub sides: Sides<Option<Smart<Rel<Length>>>>,
+    pub two_sided: Option<bool>,
+}
+
+impl Reflect for Margin {
+    fn castable(value: &Value) -> bool {
+        matches!(
+            value,
+            Value::Auto | Value::Dict(_) | Value::Length(_) | Value::Relative(_)
+        )
+    }
+
+    fn describe() -> CastInfo {
+        CastInfo::Type("auto")
+            + CastInfo::Type("dictionary")
+            + CastInfo::Type("length")
+            + CastInfo::Type("relative length")
+    }
+}
+
+impl FromValue for Margin {
+    fn from_value(value: Value) -> StrResult<Self> {
+        match value {
+            Value::Auto => Ok(Self::splat(Some(Value::cast(value)?))),
+            Value::Length(value) => Ok(Self::splat(Some(Smart::Custom(value.into())))),
+            Value::Relative(value) => Ok(Self::splat(Some(Smart::Custom(value)))),
+            Value::Dict(mut dict) => {
+                let mut take = |key| dict.take(key).ok().map(Value::cast).transpose();
+
+                let rest = take("rest")?;
+                let x = take("x")?.or(rest);
+                let y = take("y")?.or(rest);
+                let outside = take("outside")?;
+                let inside = take("inside")?;
+                let mut left = take("left")?;
+                let mut right = take("right")?;
```

Comment:
Please move `top` and `bottom` here as variables, for consistency :)

```suggestion
                let mut left = take("left")?;
                let mut right = take("right")?;
                let top =  take("top")?.or(y);
                let bottom = take("bottom")?.or(y);
```